File: templates/ros2/__subpath__/src/ros2_cpp_pkg/src/talker_node.cpp
#include "ros2_cpp_pkg/talker_node.hpp"

#include <chrono>
#include <string>

#include "ros2_cpp_pkg/greeter.hpp"

namespace ros2_cpp_pkg
{
TalkerNode::TalkerNode()
: rclcpp::Node("talker")
{
  name_ = declare_parameter<std::string>("name", "world");
  const double period_s = declare_parameter<double>("period_s", 1.0);
  pub_ = create_publisher<std_msgs::msg::String>("chatter", 10);
  timer_ = create_wall_timer(
    std::chrono::duration<double>(period_s), [this]() { tick(); });
  RCLCPP_INFO(get_logger(), "talker up -> /chatter every %.3gs", period_s);
}

void TalkerNode::tick()
{
  std_msgs::msg::String msg;
  msg.data = greeting(name_, ++count_);
  pub_->publish(msg);
}
}  // namespace ros2_cpp_pkg

File: templates/ros2/__subpath__/src/ros2_py_pkg/ros2_py_pkg/talker_node.py
"""最小 rclpy 节点：定时 publish std_msgs/String。

演示「ROS 薄壳 + 纯逻辑分层」：消息内容由纯函数 greeting() 产出（可单测），节点只管收发与定时。
"""

from __future__ import annotations

import rclpy
from rclpy.node import Node
from std_msgs.msg import String

from ros2_py_pkg.greeter import greeting


class TalkerNode(Node):
    def __init__(self) -> None:
        super().__init__("talker")
        self._name = self.declare_parameter("name", "world").value
        self._pub = self.create_publisher(String, "chatter", 10)
        self._count = 0
        period = self.declare_parameter("period_s", 1.0).value
        self.create_timer(period, self._tick)
        self.get_logger().info(f"talker up → /chatter every {period}s")

    def _tick(self) -> None:
        self._count += 1
        msg = String()
        msg.data = greeting(self._name, self._count)
        self._pub.publish(msg)


def main(args: list[str] | None = None) -> None:
    rclpy.init(args=args)
    node = TalkerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
